server/sonar-server/src/main/java/org/sonar/server/measure/ws/SearchAction.java

```diff
@@ -0,0 +1,289 @@
+        .build());
+    }
+
+    private List<String> difference(Collection<String> expected, Collection<String> actual) {
+      Set<String> actualSet = new HashSet<>(actual);
+
+      return expected.stream()
+        .filter(value -> !actualSet.contains(value))
+        .sorted(String::compareTo)
+        .collect(Collectors.toList());
+    }
+
+    private SearchWsResponse buildResponse() {
+      requireNonNull(metrics);
+      requireNonNull(measures);
+      requireNonNull(components);
+      requireNonNull(snapshots);
+
+      Map<Integer, MetricDto> metricsById = metrics.stream().collect(Collectors.toMap(MetricDto::getId, identity()));
+
+      ImmutableMultimap<String, WsMeasures.Measure> wsMeasuresByComponentUuid = Stream
+        .concat(measures.stream(), buildBestMeasures().stream())
+        .collect(Collectors.toMap(identity(), MeasureDto::getComponentUuid))
+        .entrySet().stream()
+        .map(entry -> immutableEntry(
+          measureDtoToWsMeasure(metricsById.get(entry.getKey().getMetricId()), entry.getKey()),
+          entry.getValue()))
+        .sorted((e1, e2) -> e1.getKey().getMetric().compareTo(e2.getKey().getMetric()))
+        .collect(Collector.of(
+          ImmutableMultimap::<String, WsMeasures.Measure>builder,
+          (result, entry) -> result.put(entry.getValue(), entry.getKey()),
+          (result1, result2) -> {
+            throw new IllegalStateException("Parallel execution forbidden while building WS measures");
+          },
+          ImmutableMultimap.Builder::build));
+
+      return components.stream()
+        .map(dbComponent -> dbToWsComponent(dbComponent, wsMeasuresByComponentUuid.get(dbComponent.uuid())))
+        .sorted((c1, c2) -> c1.getName().compareTo(c2.getName()))
+        .collect(Collector.of(
+          SearchWsResponse::newBuilder,
+          SearchWsResponse.Builder::addComponents,
+          (result1, result2) -> {
+            throw new IllegalStateException("Parallel execution forbidden while building SearchWsResponse");
+          },
+          SearchWsResponse.Builder::build));
+    }
+
+    private List<MeasureDto> buildBestMeasures() {
+      Set<MetricDto> metricsWithBestValue = metrics.stream()
+        .filter(metric -> metric.isOptimizedBestValue() && metric.getBestValue() != null)
+        .collect(Collectors.toSet());
+
+      Multimap<String, WsMeasures.Period> wsPeriodsByProjectUuid = snapshots.stream().collect(Collector.of(
+        ImmutableMultimap::<String, WsMeasures.Period>builder,
+        (result, snapshot) -> result.putAll(snapshot.getComponentUuid(), snapshotToWsPeriods(snapshot)),
+        (result1, result2) -> {
+          throw new IllegalStateException("Parallel execution forbidden");
+        },
+        ImmutableMultimap.Builder::build));
+
+      Table<String, Integer, MeasureDto> measuresByComponentUuidAndMetricId = measures.stream().collect(Collector.of(
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this useless assignment to local variable "measuresByComponentUuidAndMetricId". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1854)
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused "measuresByComponentUuidAndMetricId" local variable. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1481)